File: python/2368.py
from typing import List

class Solution:
    def reachableNodes(self, n: int, edges: List[List[int]], restricted: List[int]) -> int:
        g = [[] for _ in range(n)]
        for x, y in edges:
            g[x].append(y)
            g[y].append(x)
        r_s = set(restricted)
        ans = 0
        def helper(node: int, fa: int):
            nonlocal ans
            if node in r_s: return
            ans+= 1
            for nxt in g[node]:
                if nxt != fa:
                    helper(nxt, node)
        helper(0, -1)
        return ans

File: go/2368.go
package main

func reachableNodes(n int, edges [][]int, restricted []int) (ans int) {
	g := make([][]int, n)
	for _, edge := range edges {
		x, y := edge[0], edge[1]
		g[x] = append(g[x], y)
		g[y] = append(g[y], x)
	}
	m := make([]bool, n)
	for _, rest := range restricted {
		m[rest] = true
	}
	var helper func(node, fa int)
	helper = func(node, fa int) {
		if m[node] {
			return
		}
		ans++
		for _, nxt := range g[node] {
			if nxt != fa {
				helper(nxt, node)
			}
		}
	}
	helper(0, -1)
	return
}
